test/cql-pytest/test_cast_data.py

```diff
@@ -151,4 +152,53 @@ def test_cast_to_counter(cql, table3):
         with pytest.raises(InvalidRequest, match='cannot be cast'):
             cql.execute(f"SELECT CAST({col} AS counter) FROM {table3} WHERE p={p}")
 
+# Test casting 32-bit floating-point to text. The output string should have
+# enough digits to faithfully reproduce the original floating-point value
+# when read back. We don't check in this test whether maybe it outputs too
+# many digits that don't make any difference when read back.
+# Reproduces issue #15127.
+def test_cast_float_to_text(cql, table3):
+    p = unique_key_int()
+    # The numbers below (based on pi) have excessive precision, they are
+    # truncated to Python's 64-bit precision, and later further truncated
+    # to 32-bit floats when assigned to a Scylla "float" column.
+    numbers = [ 3.141592653589793238462643383279502884197,
+               -3.141592653589793238462643383279502884197,
+               -31415926535897932384.62643383279502884197,
+               0.00003141592653589793238462643383279502884197]
+    for n in numbers:
+        cql.execute(f'INSERT INTO {table3} (p, f) VALUES ({p}, {n})')
+        # Sanity check: If we read "f" back, we get n truncated to 32-bit
+        # precision. We use c_float to truncate n to 32-bit precision.
+        expected_f = c_float(n)
+        read_f = c_float(list(cql.execute(f"SELECT f FROM {table3} WHERE p={p}"))[0][0])
+        assert expected_f.value == read_f.value  # that's how to compare c_float objects
+        # Finally test CAST(f AS text): test that the resulting textual
+        # representation has enough digits to restore f without loss of
+        # precision. If we ask Scylla to cast f to text, and then convert
+        # this text back to c_float, the result should be equal again to
+        # expected_f. If the cast printed too few digits, it would not be
+        # equal.
+        ftext = list(cql.execute(f"SELECT CAST(f AS text) FROM {table3} WHERE p={p}"))[0][0]
+        read_f = c_float(float(ftext))
```

Comment:
No, both tests are important. Let me give you an example. Imagine that the Scylla fix was to change the sprintf format from "%g" to "%.8g" (for both float and double). With that "fix", the first test would pass: We will set a float (32-bit) column to whatever precision of PI it can carry, and then CAST will print it out in 8 digits of precision. The `float(ftext)` will parse this number to a double (yes, it's confusing - Python's "float" is C's and Scylla's "double"). We then trucate this to a 32-bit float (c_float). And expect that the result will be identical to the 32-bit truncation of the original number we started with. And it will, because 8 digits of precision is enough for a 32-bit float. But the second test will fail: 64-bit double ("float" type in Python) can have 15 digits of precision, so printing out only 8 of them and then scanning them will result in a different number.

You can add `print(ftext)` (and run with "cql-pytest/run --vs test_cast.py" to see this output) to see what these CASTS actually generate if you're curious. Before your Seastar patch, they generated only 6 digits of precision for both float and double types - which wasn't enough for either (the float was missing more-or-less one decimal digit of precision, double was missing much more precision).

Reply:
I think what you missed in your summar is that the two tests use a different Scylla column - one uses the float "f" and the second uses the double "db". So they exercise different Scylla "CAST ... AS text" code paths.

Reply:
ahh, true. "db" (double) versus "f" (float). sorry, i completely missed it!